### PR TITLE
script for generating pipelines README.md

### PR DESCRIPTION
**Pull request by rcerven**
[STONEBLD-2330](https://issues.redhat.com//browse/STONEBLD-2330)

# Before you complete this pull request ...

Look for any open pull requests in the repository with the title "e2e-tests update" and 
see if there are recent e2e-tests updates that will be applicable to your change.

### CODE REVIEW

**Comment by MartinBasti** [comment]
How do we want to keep it up to date? Will you add task to CI to generate readme files and compare them with HEAD?

**Comment by rcerven** [comment]
> How do we want to keep it up to date? Will you add task to CI to generate readme files and compare them with HEAD?

not sure, @arewm @mmorhun ?

File: hack/generate_help.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (rcerven) [reply hack/generate_help.py]: renamed

File: hack/generate_help.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+import re
+import yaml
+import shutil
+from pathlib import Path
+from tempfile import mkdtemp
+
+PIPELINES_DIR = './pipelines/'
+RHTAP_PIPELINES_DIR = './pipelines/rhtap/'
+TASKS_DIR = './task/'
+PIPELINE_DIFFERENT_DIR = {'gitops-pull-request': 'gitops-pull-request-rhtap'}
```
[the remark this answers is not in the document]
Comment (rcerven) [reply hack/generate_help.py]: added 1 variable for input dirs required for generation

File: hack/generate-pipelines-readme.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+                task_dict['workspaces'] = task.get('workspaces', [])
+                pipelines_info[pipeline_name]['tasks'].append(task_dict)
+
+                for param in task_dict['params']:
+                    match = param_regex.match(param['value'])
+                    if match:
+                        uses_param = match.group(1)
+                        task_param_name = f"{task_dict['name']}:{task_dict['refversion']}:{param['name']}"
+
+                        for pparam in pipelines_info[pipeline_name]['params']:
```
Comment (mmorhun) [inline hack/generate-pipelines-readme.py:105]: `pipeline_param` would be more clear.

File: hack/generate-pipelines-readme.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                                    set_by = workspace_in_pipeline['workspace']
+
+                        desc = workspace['description'].replace("\n", " ")
+                        f.write(f"|{workspace['name']}| {desc}| {workspace['optional']}| {set_by}|\n")
+
+    shutil.rmtree(temp_dir)
+
+
+if __name__ == '__main__':
+    raise SystemExit(main())
```
[the remark this answers is not in the document]
Comment (rcerven) [reply hack/generate-pipelines-readme.py:275]: will remove raising

File: hack/generate-pipelines-readme.py (hunk 3 above)
Comment (rcerven) [reply hack/generate-pipelines-readme.py:105]: renamed

File: hack/generate-pipelines-readme.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            exit(1)
+
+        pipelines_info = {}
+        with open(Path(temp_dir).joinpath(f), 'r') as f:
+            pipeline_data = yaml.safe_load(f)
+
+        pipeline_name = pipeline_data['metadata']['name']
+        pipelines_info[pipeline_name] = {'params': [], 'results': [], 'workspaces': [], 'tasks': []}
+
+        for param in pipeline_data['spec'].get('params', []):
```
[the remark this answers is not in the document]
Comment (rcerven) [reply hack/generate-pipelines-readme.py:71]: added check above to check if it is Pipeline kind

File: hack/generate-pipelines-readme.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+import re
+import yaml
+import shutil
+from pathlib import Path
+from tempfile import mkdtemp
+
+PIPELINE_GENERATE_INPUT_DIRS = ('./pipelines/', './pipelines/rhtap/')
+PIPELINES_DIR = './pipelines/'
+TASKS_DIR = './task/'
+PIPELINE_DIFFERENT_DIR = {'gitops-pull-request': 'gitops-pull-request-rhtap'}
```
[the remark this answers is not in the document]
Comment (rcerven) [reply hack/generate-pipelines-readme.py]: renamed and added comment

**Comment by mmorhun** [comment]
> How do we want to keep it up to date? Will you add task to CI to generate readme files and compare them with HEAD?

As an option we can have a PR check.
Alternatively, an action that runs periodically and updates READMEs with the script.

**Review by mmorhun — APPROVED** [review]

**Comment by MartinBasti** [comment]
> > How do we want to keep it up to date? Will you add task to CI to generate readme files and compare them with HEAD?
> 
> As an option we can have a PR check. Alternatively, an action that runs periodically and updates READMEs with the script.

That sounds nice too

File: hack/generate-pipelines-readme.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            continue
+
+        pipeline_name = pipeline_data['metadata']['name']
+        pipelines_info[pipeline_name] = {'params': [], 'results': [], 'workspaces': [], 'tasks': []}
+
+        for param in pipeline_data['spec'].get('params', []):
+            param_dict = {'used': []}
+            param_dict['name'] = param.get('name')
+            param_dict['description'] = param.get('description', "")
+            param_dict['default'] = param.get('default', None)
```
Comment (chmeliik) [inline hack/generate-pipelines-readme.py:75]: Sometimes the `default` is shown as an empty string, sometimes as "None"

image

If I understand correctly, the difference is that None means that the param is required. Can we make this more understandable for readers?

File: hack/generate-pipelines-readme.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+                                    break
+                        if set_by:
+                            set_by = f"'{set_by}'"
+
+                        desc = param['description'].replace("\n", " ")
+                        f.write(f"|{param['name']}| {desc}| {param['default']}| {set_by}|\n")
+
+                # print pipeline results
+                f.write("\n## Results\n")
+                f.write("|name|description|value|\n")
```
[the remark this answers is not in the document]
Comment (mmorhun) [reply hack/generate-pipelines-readme.py:207]: Does it make sense to have `value` column for results?

File: hack/generate-pipelines-readme.py (hunk 7 above)
Comment (rcerven) [reply hack/generate-pipelines-readme.py:75]: that isn't exactly truth .....

those params without default values are set in build service

there isn't really any real concept of required params

File: hack/generate-pipelines-readme.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rcerven) [reply hack/generate-pipelines-readme.py:207]: I guess value for results in tasks doesn't make sense as it always will be empty

[the remark this answers is not in the document]
Comment (rcerven) [reply hack/generate-pipelines-readme.py:207]: removed

File: hack/generate-pipelines-readme.py (hunk 7 above)
Comment (chmeliik) [reply hack/generate-pipelines-readme.py:75]: If a param has no default, it is required. Build-service sets the required params, but that doesn't mean they're not required

Comment (rcerven) [reply hack/generate-pipelines-readme.py:75]: already had discussion about 'required' with @arewm and we decided to not add it to create confusion

**Review by chmeliik — APPROVED** [review]

**Pull request merged**